probe-rs/src/flashing/loader.rs

```diff
@@ -198,74 +200,175 @@ impl FlashLoader {
         session: &mut Session,
         options: DownloadOptions<'_>,
     ) -> Result<(), FlashError> {
-        log::debug!("committing flash!");
+        log::debug!("committing FlashLoader!");
+
+        log::debug!("Contents of builder:");
         for (&address, data) in &self.builder.data {
             log::debug!(
-                "    region: {:08x}-{:08x} ({} bytes)",
+                "    data: {:08x}-{:08x} ({} bytes)",
                 address,
                 address + data.len() as u32,
                 data.len()
             );
         }
 
+        log::debug!("Flash algorithms:");
+        for algorithm in &session.target().flash_algorithms {
+            let Range { start, end } = algorithm.flash_properties.address_range;
+
+            log::debug!(
+                "    algo {}: {:08x}-{:08x} ({} bytes)",
+                algorithm.name,
+                start,
+                end,
+                end - start
+            );
+        }
+
         // Iterate over all memory regions, and program their data.
 
         if self.memory_map != session.target().memory_map {
             log::warn!("Memory map of flash loader does not match memory map of target!");
         }
 
+        let mut algos: HashMap<String, Vec<NvmRegion>> = HashMap::new();
+
         // Commit NVM first
+
+        // Iterate all NvmRegions and group them by flash algorithm.
+        // This avoids loading the same algorithm twice if it's used for two regions.
+        //
+        // This also ensures correct operation when chip erase is used. We assume doing a chip erase
+        // using a given algorithm erases all regions controlled by it. Therefore, we must do
+        // chip erase once per algorithm, not once per region. Otherwise subsequent chip erases will
+        // erase previous regions' flashed contents.
+        log::debug!("Regions:");
         for region in &self.memory_map {
             if let MemoryRegion::Nvm(region) = region {
-                self.commit_nvm(region, session, &options)?;
+                log::debug!(
+                    "    region: {:08x}-{:08x} ({} bytes)",
+                    region.range.start,
+                    region.range.end,
+                    region.range.end - region.range.start
+                );
+
+                // If we have no data in this region, ignore it.
+                // This avoids uselessly initializing and deinitializing its flash algorithm.
+                if !self.builder.has_data_in_range(&region.range) {
+                    log::debug!("     -- empty, ignoring!");
+                    continue;
+                }
+
+                let algo = Self::get_flash_algorithm_for_region(region, session.target())?;
+                if !algos.contains_key(&algo.name) {
+                    algos.insert(algo.name.clone(), Vec::new());
+                }
+
+                algos.get_mut(&algo.name).unwrap().push(region.clone());
```

Comment:
I like to use the `entry` function for this, to avoid the unwrap:

```rust
let entry = algos.entry(algo.name.clone()).or_default();

entry.push(region.clone());
```

But I'm also happy with the current implementation ;)

Reply:
that's a way cooler way of doing it, indeed! Changing